src/config.rs

```diff
@@ -0,0 +1,178 @@
+
+/// Represents the configuration for the Sprocket `format` command.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case", deny_unknown_fields)]
+pub struct FormatConfig {
+    /// Use tabs for indentation (default is spaces).
+    pub with_tabs: bool,
+    /// The number of spaces to use for indentation levels (default is 4).
+    pub indentation_size: usize,
+    /// The maximum line length (default is 90).
+    pub max_line_length: usize,
+}
+
+impl Default for FormatConfig {
+    fn default() -> Self {
+        Self {
+            with_tabs: false,
+            indentation_size: 4,
+            max_line_length: 90,
+        }
+    }
+}
+
+/// Represents the configuration for the Sprocket `check` and `lint` commands.
+#[derive(Debug, Clone, Default, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case", deny_unknown_fields)]
+pub struct CheckConfig {
+    /// Rule IDs to except from running.
+    pub except: Vec<String>,
+    /// Causes the command to fail if any warnings are reported.
+    pub deny_warnings: bool,
+    /// Causes the command to fail if any notes are reported.
+    pub deny_notes: bool,
+    /// Hide diagnostics with `note` severity.
+    pub hide_notes: bool,
+}
+
+/// Represents the configuration for the Sprocket `run` command.
+#[derive(Debug, Clone, Default, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case", deny_unknown_fields)]
+pub struct RunConfig {
+    /// Path to the engine configuration file.
+    pub config: PathBuf,
+}
+
+impl Config {
+    /// Create a new config instance by reading potential configurations.
+    pub fn new(path: Option<String>) -> Self {
+        // Check for a config file in the current directory
+        // Start a new Figment instance with default values
+        let mut figment = Figment::new().admerge(Serialized::from(Config::default(), "default"));
+
+        // Check XDG_CONFIG_HOME for a config file
+        // On MacOS, check HOME for a config file
+        #[cfg(target_os = "macos")]
+        {
+            if let Some(home) = dirs::home_dir() {
+                tracing::info!(
+                    "reading configuration from: {home:?}/.config/sprocket/sprocket.toml"
+                );
+                figment = figment.admerge(Toml::file(
+                    home.join(".config").join("sprocket").join("sprocket.toml"),
+                ));
+            }
+        }
+        #[cfg(not(target_os = "macos"))]
+        {
+            if let Some(xdg_config_home) = dirs::config_dir() {
+                tracing::info!(
+                    "reading configuration from XDG_CONFIG_HOME: \
+                     {xdg_config_home:?}/sprocket/sprocket.toml"
+                );
+                figment = figment.admerge(Toml::file(
+                    xdg_config_home.join("sprocket").join("sprocket.toml"),
+                ));
+            }
+        }
+
+        // Check PWD for a config file
+        if Path::exists(Path::new("sprocket.toml")) {
```

Comment:
Big question: should it be `sprocket.toml` or `Sprocket.toml`. I think the latter personally in all cases.

Reply:
I disagree. I do not like capitalized file names generally. Without doing any investigation, I feel like most configuration files are lowercased (with `Cargo.toml` being a notable exception).

Reply:
I could go either way, with a slight preference for `sprocket.toml`.

Fwiw, Diesel, a popular Rust ORM, uses lowercase `diesel.toml`.

Reply:
> I could go either way, with a slight preference for `sprocket.toml`.

same here